common/multigrid/amgx_pgm_kernels.hpp.inc

```diff
@@ -0,0 +1,327 @@
+        }
+    }
+    if (strongest_agg != -1) {
+        agg_val[row] = agg_const_val[strongest_agg];
+    } else {
+        agg_val[row] = row;
+    }
+}
+
+// This is the undeterminstic implementation which is the same implementation of
+// the previous one but agg_val == agg_const_val.
+template <typename ValueType, typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void assign_to_exist_agg_kernel(
+        const size_type num, const IndexType *__restrict__ row_ptrs,
+        const IndexType *__restrict__ col_idxs,
+        const ValueType *__restrict__ weight_vals,
+        const ValueType *__restrict__ diag, IndexType *__restrict__ agg_val)
+{
+    auto row = thread::get_thread_id_flat();
+    if (row >= num || agg_val[row] != -1) {
+        return;
+    }
+    ValueType max_weight_agg = zero<ValueType>();
+    IndexType strongest_agg = -1;
+    for (auto idx = row_ptrs[row]; idx < row_ptrs[row + 1]; idx++) {
+        auto col = col_idxs[idx];
+        if (col == row) {
+            continue;
+        }
+        auto weight = weight_vals[idx] / max(abs(diag[row]), abs(diag[col]));
+        if (agg_val[col] != -1 &&
+            thrust::tie(weight, col) >
+                thrust::tie(max_weight_agg, strongest_agg)) {
+            max_weight_agg = weight;
+            strongest_agg = col;
+        }
+    }
+    if (strongest_agg != -1) {
+        agg_val[row] = agg_val[strongest_agg];
+    } else {
+        agg_val[row] = row;
+    }
+}
+
+
+template <typename IndexType>
+__global__ __launch_bounds__(default_block_size) void get_source_row_map_kernel(
+    const size_type source_nrows, const IndexType *__restrict__ agg_val,
+    const IndexType *__restrict__ source_row_ptrs,
+    IndexType *__restrict__ result_row_ptrs, IndexType *__restrict__ row_map)
+{
+    auto row = thread::get_thread_id_flat();
+    if (row >= source_nrows) {
+        return;
+    }
+    const auto num_elems = source_row_ptrs[row + 1] - source_row_ptrs[row];
+    const auto result_idx = agg_val[row];
+    // atomic_add returns the old value, so it can be the starting point.
+    row_map[row] = atomic_add(result_row_ptrs + result_idx, num_elems);
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void move_row_kernel(
+    const size_type source_nrows, const IndexType *__restrict__ agg_val,
+    const IndexType *__restrict__ row_map,
+    const IndexType *__restrict__ source_row_ptrs,
+    const IndexType *__restrict__ source_col_idxs,
+    const ValueType *__restrict__ source_values,
+    const IndexType *__restrict__ result_row_ptrs,
+    IndexType *__restrict__ result_col_idxs,
+    ValueType *__restrict__ result_values)
+{
+    auto row = thread::get_thread_id_flat();
+    if (row >= source_nrows) {
+        return;
+    }
+    auto result_i = result_row_ptrs[agg_val[row]] + row_map[row];
+    for (auto i = source_row_ptrs[row]; i < source_row_ptrs[row + 1];
```

Comment:
small nit: could you move result_i++ to the loop body? Otherwise this might be a bit hidden/unexpected to the user, as below with copy_to_coarse_kernel